src/autoscaler/metricsgateway/helpers/ws_helper.go

```diff
@@ -82,75 +109,66 @@ func (wh *wshelper) SetupConn() error {
 					}
 				}
 			}()
-			wh.lock.Lock()
-			defer wh.lock.Unlock()
-			wh.wsConn = con
+			wh.connection.setConnection(con)
 			return nil
 		}
 	}
 }
 
-func (wh *wshelper) CloseConn() error {
-	retryCount := 1
+func (wh *WsHelper) IsClosed() bool {
+	return wh.connection.getConnection() == nil
+}
+
+func (wh *WsHelper) CloseConn() error {
+	retryCount := 0
 	for {
 		wh.logger.Info("close-ws-connection")
-		err := wh.wsConn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""), time.Time{})
+		err := wh.connection.getConnection().WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""), time.Time{})
 		if err != nil {
-			wh.logger.Error("failed-to-send-close-message-to-metricserver", err, lager.Data{"current": retryCount})
-			if retryCount <= wh.maxCloseRetryCount {
+			if retryCount < wh.maxCloseRetryCount {
 				retryCount++
-				wh.logger.Info("retry", lager.Data{"RETRY": retryCount})
 				time.Sleep(wh.retryDelay)
 			} else {
-				wh.logger.Error("maximum-number-of-close-retries-reached", err, lager.Data{"maxCloseRetryCount": wh.maxCloseRetryCount})
-				return err
+				return fmt.Errorf("failed to close correctly after %d retries: %w", retryCount, err)
 			}
 		} else {
 			go func() {
-				wh.lock.Lock()
-				con := wh.wsConn
-				wh.lock.Unlock()
-				time.AfterFunc(5*time.Second, func() {
-					err := con.Close()
-					if err != nil {
-						wh.logger.Error("failed-to-close-ws-connection", err)
-					} else {
-						wh.logger.Info("successfully-close-ws-connection")
-					}
+				time.AfterFunc(wh.CloseWaitTime, func() {
+					_ = wh.connection.Close()
 				})
 			}()
 			return nil
 		}
 	}
 }
```

Comment:
If a Write() fails we try to `reconnect` which in happy path is just
```
	wh.CloseConn()
	wh.SetupConn()
```
if `wh.connection.getConnection().WriteControl(websocket.CloseMessage, ...)` succeeds we schedule a `wh.connection.Close()` and return.

Inside that time `wh.SetupConn()` could setup a new connection which would immediately be closed.

I have the feeling either the whole `CloseCon` needs to be in the write lock/be part of connection or, maybe more appropriate, the goroutine should have a private copy of the connection it then closes and immediately set the `wh.connection` to `nil`.